Thermally coupled Constitutive Laws

AlejandroCornejo opened the pull request:
**Description**
Dear @KratosMultiphysics/structural-mechanics and @KratosMultiphysics/constitutive-models 

I'm implementing all the code infraestructure in order to be able to simulate **thermo-mechanical calculations** with a solver that couples the `Structural` and the  `ConvectionDiffusion` Apps.

The solver was already there long time ago, I've just improved it a bit and corrected a detail.


The setting of the `REFERENCE_TEMPERATURE`, different options are implemented: using the `GetValue()` in the `Geometry` (has to be set outside by someone) and being a material property as:
~~~cpp
void ThermalLinearPlaneStrain::InitializeMaterial(
    const Properties& rMaterialProperties,
    const GeometryType& rElementGeometry,
    const Vector& rShapeFunctionsValues
    )
{
    if (rElementGeometry.Has(REFERENCE_TEMPERATURE)) {
        mReferenceTemperature = rElementGeometry.GetValue(REFERENCE_TEMPERATURE);
    } else if (rMaterialProperties.Has(REFERENCE_TEMPERATURE)) {
        mReferenceTemperature = rMaterialProperties[REFERENCE_TEMPERATURE];
    }
}
~~~
**Changelog**
Please summarize the changes in one list to generate the changelog:
E.g.
- Erased the convective term in the Eulerian calculation
- Creation of the new thermal CL inside the ConstitutiveLawsApp, only the elastic basic ones
- etc...

**Documentation**

Now, the `materials.json` for the structural problem looks as this:

~~~json
{
    "properties" : [{
        "model_part_name" : "Structure.Parts_Solid_Solid_Auto1",
        "properties_id"   : 1,
        "Material"        : {
            "constitutive_law" : {
                "name" : "ThermalLinearPlaneStrain"
                
            },
            "Variables"        : {
                "DENSITY"       : 7850.0,
                "YOUNG_MODULUS" : 2.1e11,
                "POISSON_RATIO" : 0.29,
                "THERMAL_EXPANSION_COEFFICIENT" : 7.2e-6,
                "REFERENCE_TEMPERATURE" : 22.5
            },
            "Tables"           : {
                "TEMPERATURE_vs_E" : {
                        "input_variable"  : "TEMPERATURE",
                        "output_variable" : "YOUNG_MODULUS",
                        "data"            : [[-1,   2.1e11],
                                             [3500, 2.1e8 ],
                                             [1e6,  2.1e8 ]]
                    }
            },
        "accessors"        : {
            "accessor_table_T_E" : {
                "accessor_type"  : "table_accessor", // here we indicate the table accessor
                "properties" : {
                    "table_input_variable"      : "TEMPERATURE",
                    "table_output_variable"     : "YOUNG_MODULUS",
                    "table_input_variable_type" : "node_historical" // we support several types

				}
            }
        }
        }
    }]
}
~~~

AlejandroCornejo committed:
minor

AlejandroCornejo committed:
clear print

AlejandroCornejo committed:
adding convective correctiion

loumalouomega commented on applications/ConvectionDiffusionApplication/python_scripts/coupled_structural_thermal_solver.py:
```diff
@@ -149,11 +150,23 @@ def Predict(self):
+
+    def RemoveConvectiveVelocity(self):
+        if self.is_dynamic:
+            for node in self.thermal_solver.GetComputingModelPart().Nodes:
```
USe VariableUtils, is faster

AlejandroCornejo replied:
Yes!! this was provisional. Now that you've mentioned, the variable utils can assign to the mesh velocity the velocity???? I was thinking in implementing the utility from scratch

loumalouomega replied:
There are methods, let me check the name in python

loumalouomega replied:
CopyModelPartNodalVar

AlejandroCornejo replied:
Checking... THX

loumalouomega replied:
```suggestion
KratosMultiphysics.VariableUtils().CopyModelPartNodalVar(KratosMultiphysics.VELOCITY, KratosMultiphysics.MESH_VELOCITY, self.thermal_solver.GetComputingModelPart(), self.thermal_solver.GetComputingModelPart(), 0)
            
```

loumalouomega commented on applications/ConvectionDiffusionApplication/python_scripts/coupled_structural_thermal_solver.py:
```diff
@@ -149,11 +150,23 @@ def Predict(self):
         thermal_is_converged = self.thermal_solver.SolveSolutionStep()
 
-        return (fluid_is_converged and thermal_is_converged)
+        print("+--------------------------------------------------------------+")
```
Please, use logger 
```suggestion
        KratosMultiphysics.Logger.PrintInfo("+--------------------------------------------------------------+")
```

loumalouomega committed:
Update applications/ConvectionDiffusionApplication/python_scripts/coupled_structural_thermal_solver.py

loumalouomega reviewed (changes requested):
Se my comment of VariableUtils

AlejandroCornejo committed:
print leftover

AlejandroCornejo committed:
using variable_utils for convective velocity

AlejandroCornejo committed:
adding generic virtual method in base solid CL in SMA

loumalouomega commented on applications/ConvectionDiffusionApplication/python_scripts/coupled_structural_thermal_solver.py:
```diff
@@ -149,11 +150,23 @@ def Predict(self):
+    def RemoveConvectiveVelocity(self):
+        if self.is_dynamic:
+            for node in self.thermal_solver.GetComputingModelPart().Nodes:
+                node.SetSolutionStepValue(KratosMultiphysics.MESH_VELOCITY, node.GetSolutionStepValue(KratosMultiphysics.VELOCITY))
```
```suggestion
```

loumalouomega committed:
VariableUtils

AlejandroCornejo committed:
Merge branch 'thermal-solver-mechanical-coupling' of https://github.com/KratosMultiphysics/Kratos into thermal-solver-mechanical-coupling

loumalouomega commented on applications/ConvectionDiffusionApplication/python_scripts/coupled_structural_thermal_solver.py:
```diff
@@ -149,11 +150,23 @@ def Predict(self):
+
+    def RemoveConvectiveVelocity(self):
+        if self.is_dynamic:
+KratosMultiphysics.VariableUtils().CopyModelPartNodalVar(KratosMultiphysics.VELOCITY, KratosMultiphysics.MESH_VELOCITY, self.thermal_solver.GetComputingModelPart(), self.thermal_solver.GetComputingModelPart(), 0)
```
```suggestion
            KratosMultiphysics.VariableUtils().CopyModelPartNodalVar(KratosMultiphysics.VELOCITY, KratosMultiphysics.MESH_VELOCITY, self.thermal_solver.GetComputingModelPart(), self.thermal_solver.GetComputingModelPart(), 0)
```

loumalouomega committed:
Tabs

AlejandroCornejo committed:
merge

philbucher commented on applications/ConvectionDiffusionApplication/python_scripts/coupled_structural_thermal_solver.py:
```diff
@@ -67,6 +67,7 @@ def __init__(self, model, custom_settings):
 
         from KratosMultiphysics.ConvectionDiffusionApplication import python_solvers_wrapper_convection_diffusion
         self.thermal_solver = python_solvers_wrapper_convection_diffusion.CreateSolverByParameters(self.model,self.settings["thermal_solver_settings"],"OpenMP")
+        self.is_dynamic = self.settings["structural_solver_settings"]["solver_type"].GetString() == "Dynamic"
```
Lowercase is also valid input (and the preferred one)

AlejandroCornejo committed:
accepting lowercase

AlejandroCornejo committed:
elastic 3d law modified

AlejandroCornejo committed:
unused mat_props

AlejandroCornejo committed:
changinh plane strain

AlejandroCornejo committed:
clear spaces

AlejandroCornejo committed:
plane stress

AlejandroCornejo committed:
axysimmetric

philbucher commented on applications/StructuralMechanicsApplication/custom_constitutive/linear_plane_stress.cpp:
```diff
@@ -117,9 +116,8 @@ void LinearPlaneStress::CalculatePK2Stress(
-    const double E = r_material_properties[YOUNG_MODULUS];
-    const double NU = r_material_properties[POISSON_RATIO];
+    const double E  = GetMaterialProperty(YOUNG_MODULUS, rValues);
+    const double NU = GetMaterialProperty(POISSON_RATIO, rValues);
```
what about the elements?
=> what if they use `r_material_properties[YOUNG_MODULUS]`?

AlejandroCornejo replied:
hmmm should they? I'm thinking about standard solid FE. In what cases this can happen?

philbucher replied:
I am not sure
a quick search showed that it is not used often but sometimes
Can you take a look yourself at those cases?

philbucher commented on applications/StructuralMechanicsApplication/custom_constitutive/elastic_isotropic_3d.h:
```diff
@@ -360,6 +360,15 @@ class KRATOS_API(STRUCTURAL_MECHANICS_APPLICATION) ElasticIsotropic3D
+        )
+    {
+        return rValues.GetMaterialProperties()[rVariable];
+    }
```
@RiccardoRossi do you agree with making this virtual?

philbucher replied:
If we allow this, then it could also be a semi-solution for #2349 (one would still need a derived and specialized CLaw but better than nothing)

So if the performance penalty is not too large I am positive about adding this!

philbucher reviewed (changes requested):
can you please make some performance tests for adding this virtual method?
This is called a lot of times

Other than that the changes look ok to me, but see my question about the access of the props from the elements
(which was one of the main reasons #2349 failed)

AlejandroCornejo committed:
adding table interpolation in advanced constitutive laws utilities

AlejandroCornejo committed:
adding comment in method

AlejandroCornejo committed:
inverting order -> Reference TEMPERATURE

AlejandroCornejo committed:
more methods to the utils

AlejandroCornejo committed:
adding first thermal elastic 3d law

AlejandroCornejo committed:
adding calculation methods

AlejandroCornejo committed:
registering thermal law

AlejandroCornejo committed:
minor strain type

AlejandroCornejo committed:
final details of the first thermal law

AlejandroCornejo committed:
better looking logger

AlejandroCornejo committed:
clang issue

AlejandroCornejo committed:
registering plane strain

AlejandroCornejo committed:
minor original plane strain space

AlejandroCornejo committed:
adding plane strain thermal

AlejandroCornejo committed:
consistent ordering

AlejandroCornejo committed:
correct ordering of Solve in the Thermal and Structural

now the predict is not altered by the old Thermal strain, it uses the most updated one

philbucher commented on applications/ConstitutiveLawsApplication/custom_constitutive_thermal/thermal_elastic_isotropic_3d.cpp:
```diff
@@ -0,0 +1,151 @@
+    KRATOS_TRY
+    const auto& r_process_info = rValues.GetProcessInfo();
+    if (r_process_info[STEP] == 1) // We storage the ref temperature as the initial one
+        mReferenceTemperature = AdvancedConstitutiveLawUtilities<6>::CalculateInGaussPoint(TEMPERATURE, rValues);
```
this should be done in `Initialize` (and check for restart)

loumalouomega replied:
I agree with @philbucher

AlejandroCornejo replied:
Yeah I thought about doing that. It is a solution depending on your needs. In this case I adopted that the ref Temp was the one at the very beginning of the calculation (can be different FE wise). In this case, **it is required that the Thermal solver at leas has computed the Temp field once**, and this is after the `Initialize `method I think.

AlejandroCornejo replied:
The reference Temp could be also a material property BTW

loumalouomega replied:
Can be a property... I am thinking that you maybe can define the temp information as a pointer to a struct, and check if the pointer is null (so if it is not possible to initialize in the Initialize can be checked later)... I don't know, there are many options. Can be just simply be set with the "SetValue" method, and be set outside, from the reading processes or some other point of the solver. There are many options

loumalouomega replied:
Opinions @philbucher ? (in any case  the SetValue should be implemented)

philbucher replied:
I am not sure if we need such a complicated solution for the ref temp
I think having `REFERENCE_TEMPERATURE` would be sufficient
```c++
// in Initialize
if properties.Has(REFERENCE_TEMPERATURE)
    mReferenceTemperature = REFERENCE_TEMPERATURE;
else
    mReferenceTemperature = TEMPERATURE;
```

or is that too simple?

AlejandroCornejo replied:
yes, I'm going to implement the option with the SetValue

AlejandroCornejo replied:
@philbucher This will be the simplest option, but if we want a ref temp element-wise, we should add as many material properties as FE! :D

philbucher replied:
ah
in this case I suggest:
```c++
// in Initialize
if this->Has(REFERENCE_TEMPERATURE)
    mReferenceTemperature = REFERENCE_TEMPERATURE;
else if properties.Has(REFERENCE_TEMPERATURE)
    mReferenceTemperature = REFERENCE_TEMPERATURE;
else
    mReferenceTemperature = TEMPERATURE;
```

AlejandroCornejo replied:
exactly, this is what I'm going to implement. The last option is the problematic in this case because the TEMPERATURE is not computed yet (Initialize state). I'll try to circumvent this problem

loumalouomega replied:
We can add a new flag ... (opening Pandora's box) 

TEMPERATURE_INITIALIZED for example

loumalouomega replied:
> We can add a new flag ... (opening Pandora's box)
> 
> TEMPERATURE_INITIALIZED for example

I see my proposal is not welcome. We can just simply modify the reading material process, so this is set on that reading fase... In any case ould be troublesome

adityaghantasala commented on applications/ConstitutiveLawsApplication/custom_constitutive_thermal/thermal_elastic_isotropic_3d.cpp:
```diff
@@ -0,0 +1,151 @@
+{
+    KRATOS_TRY
+    const auto& r_process_info = rValues.GetProcessInfo();
+    if (r_process_info[STEP] == 1) // We storage the ref temperature as the initial one
```
I actually do not like this way of taking the reference temperature. 
What if we restart and from a different ref temperature ?
Can't we make a new variable or use an existing variable to set this reference temperature ?

adityaghantasala replied:
Please correct me if I am wrong, this function is called for every element no ? 
So this check if STEP==1 is done for each element right ? This looks expensive !

AlejandroCornejo replied:
Definitely, as I said at the beginning this is a work in progress PR :D 
In this case I adopted that the ref Temp was the one at the very beginning of the calculation, would you prefer to set it via material property?

AlejandroCornejo replied:
This is now solved

loumalouomega commented on applications/ConstitutiveLawsApplication/custom_constitutive_thermal/thermal_elastic_isotropic_3d.cpp:
```diff
@@ -0,0 +1,151 @@
+        r_flags.Set(ConstitutiveLaw::COMPUTE_STRESS, flag_stress);
+    } else if (rThisVariable == INITIAL_STRAIN_VECTOR) {
+        if (this->HasInitialState()) {
+	    if (rValue.size() != GetStrainSize()) {
```
Take care of the mixing of tabs and spaces

AlejandroCornejo replied:
solved

loumalouomega commented on applications/ConstitutiveLawsApplication/custom_utilities/advanced_constitutive_law_utilities.h:
```diff
@@ -478,6 +479,75 @@ class KRATOS_API(CONSTITUTIVE_LAWS_APPLICATION) AdvancedConstitutiveLawUtilities
         BoundedMatrixVoigtType &rNewOperator
     );
 
+    /**
```
Test for this method will be nice

AlejandroCornejo replied:
yes, at the end of the PR i'll add tests for everything

loumalouomega commented on applications/ConstitutiveLawsApplication/custom_utilities/advanced_constitutive_law_utilities.h:
```diff
@@ -478,6 +479,75 @@ class KRATOS_API(CONSTITUTIVE_LAWS_APPLICATION) AdvancedConstitutiveLawUtilities
+
+        // Compute the independent variable at the Gauss point
+        double independent_at_gauss = 0.0;
+        for (unsigned int i = 0; i < r_N.size(); ++i) {
```
```suggestion
        for (std::size_t i = 0; i < r_N.size(); ++i) {
```

loumalouomega commented on applications/ConstitutiveLawsApplication/custom_utilities/advanced_constitutive_law_utilities.h:
```diff
@@ -478,6 +479,75 @@ class KRATOS_API(CONSTITUTIVE_LAWS_APPLICATION) AdvancedConstitutiveLawUtilities
+    {
+
+        const GeometryType& r_geometry = rParameters.GetElementGeometry();
+        const unsigned int number_of_nodes = r_geometry.size();
```
```suggestion
        const std::size_t number_of_nodes = r_geometry.size();
```

loumalouomega reviewed (changes requested):
More comments

AlejandroCornejo committed:
Update applications/ConstitutiveLawsApplication/custom_utilities/advanced_constitutive_law_utilities.h

Co-authored-by: Vicente Mataix Ferrándiz <vmataix@cimne.upc.edu>

AlejandroCornejo committed:
Update applications/ConstitutiveLawsApplication/custom_utilities/advanced_constitutive_law_utilities.h

Co-authored-by: Vicente Mataix Ferrándiz <vmataix@cimne.upc.edu>

AlejandroCornejo committed:
adding csome calculate values

rubenzorrilla commented on applications/ConvectionDiffusionApplication/python_scripts/coupled_structural_thermal_solver.py:
```diff
@@ -67,6 +67,8 @@ def __init__(self, model, custom_settings):
         from KratosMultiphysics.ConvectionDiffusionApplication import python_solvers_wrapper_convection_diffusion
         self.thermal_solver = python_solvers_wrapper_convection_diffusion.CreateSolverByParameters(self.model,self.settings["thermal_solver_settings"],"OpenMP")
+        solver_type = self.settings["structural_solver_settings"]["solver_type"].GetString()
+        self.is_dynamic = solver_type == "Dynamic" or solver_type == "dynamic"
```
```suggestion
        self.is_dynamic = solver_type == "dynamic"
```
I wouldn't support old input syntax in new solvers.

philbucher replied:
can you pleas change this?

AlejandroCornejo committed:
adding custom utilities to python in CLApp

AlejandroCornejo committed:
adding blanck utility

AlejandroCornejo committed:
imprving the solver

AlejandroCornejo committed:
not required in the end the python process

AlejandroCornejo committed:
solving REFERENCE_TEMP in CLs

AlejandroCornejo committed:
adding empty utilities container for the future

AlejandroCornejo committed:
registering thermal plane zstress

AlejandroCornejo committed:
plane stress thermal cl

AlejandroCornejo committed:
missing overrides

AlejandroCornejo committed:
avoiding warning

philbucher commented on applications/ConvectionDiffusionApplication/python_scripts/coupled_structural_thermal_solver.py:
```diff
@@ -7,6 +7,7 @@
 # Import applications
 import KratosMultiphysics.StructuralMechanicsApplication as KratosSMA
 import KratosMultiphysics.ConvectionDiffusionApplication as ConvDiff
+import KratosMultiphysics.ConstitutiveLawsApplication    as KratosCLApp
```
is this always necessary?
Cannot it be used without too?

AlejandroCornejo replied:
sorry it's a leftover, rrrremoving it :D

AlejandroCornejo committed:
unused import

loumalouomega committed:
Methods are access (not operators)

loumalouomega committed:
Minor

AlejandroCornejo committed:
adding c++ test, 2d version

AlejandroCornejo committed:
unused includes

AlejandroCornejo committed:
adding test with table interpolation

AlejandroCornejo committed:
adding 3d case thermal with tables interpolation

AlejandroCornejo committed:
minor for checks

AlejandroCornejo committed:
removing virtual from utilities

AlejandroCornejo commented on applications/StructuralMechanicsApplication/custom_constitutive/elastic_isotropic_3d.h:
```diff
@@ -360,6 +360,21 @@ class KRATOS_API(STRUCTURAL_MECHANICS_APPLICATION) ElasticIsotropic3D
+        ConstitutiveLaw::Parameters& rValues
+        )
+    {
+        return rValues.GetMaterialProperties()[rVariable];
```
sorry! there is no `if`, only the virtualization of this method!! @philbucher

loumalouomega replied:
That's why we can do it with an if constexpr in C++17... (or we can even do it in C++11 with some https://en.cppreference.com/w/cpp/types/conditional and some magic lambda, do you want to try it?)

philbucher replied:
@AlejandroCornejo ah ok thats also what I had in mind

@loumalouomega I don't see what exactly you want to do
if you try can you please make a derived branch? I am already losing the overview here :sweat_smile:

AlejandroCornejo replied:
Yes, I've talked with @loumalouomega about this, we will make improvements in another branch

loumalouomega replied:
> That's why we can do it with an if constexpr in C++17... (or we can even do it in C++11 with some [en.cppreference.com/w/cpp/types/conditional](https://en.cppreference.com/w/cpp/types/conditional) and some magic lambda, do you want to try it?)

Lambda and std::function are similar in nature, but in order to implement requires a little extra of work, so I would do it in a independent PR (or I can dedicate some minutes to this)

ddiezrod commented on applications/ConstitutiveLawsApplication/custom_utilities/advanced_constitutive_law_utilities.h:
```diff
@@ -478,6 +479,86 @@ class KRATOS_API(CONSTITUTIVE_LAWS_APPLICATION) AdvancedConstitutiveLawUtilities
+        ConstitutiveLaw::Parameters& rParameters
+        )
+    {
+        double alpha = rParameters.GetMaterialProperties()[THERMAL_EXPANSION_COEFFICIENT];
```
In our experience in casting, this one is also temperature dependent...

AlejandroCornejo replied:
Interesting, in order to not increase the already big PR, I'll add this in the future if you do not mind :D

rubenzorrilla replied:
Considering that `Properties` support tables, this shouldn't be a problem right?

AlejandroCornejo replied:
not at all... I can do it in this PR if you think that it is important

philbucher replied:
this needs the same solution with the virtual fct no?

rubenzorrilla replied:
Maybe we can leave it for a  future PR but I agree @ddiezrod that it should be possible to make these kind of material properties temperature dependent. Hence we'll need tables for sure.

AlejandroCornejo replied:
I'll do it this afternoon or tomorrow

AlejandroCornejo replied:
done :D

AlejandroCornejo committed:
adding temperature table possibility for alpha

loumalouomega committed:
Merge branch 'master' into thermal-solver-mechanical-coupling

loumalouomega committed:
[WIP] Function example

AlejandroCornejo committed:
Merge branch 'master' into thermal-solver-mechanical-coupling

AlejandroCornejo committed:
Merge branch 'master' into thermal-solver-mechanical-coupling

AlejandroCornejo committed:
Merge branch 'master' into thermal-solver-mechanical-coupling

AlejandroCornejo committed:
Merge branch 'master' into thermal-solver-mechanical-coupling

AlejandroCornejo committed:
Merge branch 'master' into thermal-solver-mechanical-coupling

AlejandroCornejo committed:
Merge branch 'master' into thermal-solver-mechanical-coupling

AlejandroCornejo committed:
Merge branch 'master' into thermal-solver-mechanical-coupling

AlejandroCornejo committed:
Merge branch 'master' into thermal-solver-mechanical-coupling

AlejandroCornejo committed:
Merge branch 'master' into thermal-solver-mechanical-coupling

AlejandroCornejo committed:
Merge branch 'master' into thermal-solver-mechanical-coupling

AlejandroCornejo committed:
updating Check method

AlejandroCornejo committed:
Merge branch 'master' into thermal-solver-mechanical-coupling

AlejandroCornejo committed:
Merge branch 'master' into thermal-solver-mechanical-coupling

AlejandroCornejo committed:
Merge branch 'master' into thermal-solver-mechanical-coupling

AlejandroCornejo committed:
Merge branch 'master' into thermal-solver-mechanical-coupling

AlejandroCornejo committed:
Merge branch 'master' into thermal-solver-mechanical-coupling

AlejandroCornejo committed:
Merge branch 'master' into thermal-solver-mechanical-coupling

AlejandroCornejo committed:
Merge branch 'master' into thermal-solver-mechanical-coupling

AlejandroCornejo committed:
Merge branch 'master' into thermal-solver-mechanical-coupling

AlejandroCornejo committed:
Merge branch 'master' into thermal-solver-mechanical-coupling

AlejandroCornejo committed:
Merge branch 'master' into thermal-solver-mechanical-coupling

AlejandroCornejo committed:
Merge branch 'master' into thermal-solver-mechanical-coupling

AlejandroCornejo committed:
Merge branch 'master' into thermal-solver-mechanical-coupling

AlejandroCornejo committed:
Merge branch 'master' into thermal-solver-mechanical-coupling

AlejandroCornejo committed:
Merge branch 'master' into thermal-solver-mechanical-coupling

AlejandroCornejo committed:
Merge branch 'master' into thermal-solver-mechanical-coupling

AlejandroCornejo committed:
unused print

AlejandroCornejo committed:
adding removed methods

AlejandroCornejo committed:
ne more

AlejandroCornejo committed:
separated in cpp version

AlejandroCornejo committed:
minor

AlejandroCornejo committed:
Merge branch 'master' into thermal-solver-mechanical-coupling

AlejandroCornejo committed:
Merge branch 'master' into thermal-solver-mechanical-coupling

AlejandroCornejo committed:
revert overhead in elastic ones

AlejandroCornejo committed:
no longer used this way

AlejandroCornejo committed:
missing props

AlejandroCornejo committed:
exposing E and NU to be able to use accessor dependent props

AlejandroCornejo committed:
update damage

AlejandroCornejo committed:
exposing properties and moving to utilities the calculation of PK2 stress from strain

AlejandroCornejo committed:
updating Thermal elastic 3D

AlejandroCornejo committed:
minor tab

AlejandroCornejo committed:
adapting plane stress

AlejandroCornejo committed:
plane strain ok

AlejandroCornejo committed:
minor correction in plane strain

AlejandroCornejo committed:
remove unused method and file

AlejandroCornejo committed:
remove include

AlejandroCornejo committed:
movign thermal CL

AlejandroCornejo committed:
more mving

AlejandroCornejo committed:
cpp tests updated

philbucher commented on applications/ConvectionDiffusionApplication/python_scripts/coupled_structural_thermal_solver.py:
```diff
@@ -141,19 +143,30 @@ def AdvanceInTime(self, current_time):
+        self.thermal_solver.Predict()
+
+        KratosMultiphysics.Logger.PrintInfo("\n" + "\t" + "Solving Thermal part..." + "\n")
+        self.thermal_solver.SolveSolutionStep()
```
this step should also be done in `SolveSolutionStep`

AlejandroCornejo replied:
I am doing it in this way since the predict of the mechanical part should be with the updated temperature, otherwise we cannot "converge" since the residual was badly computed

philbucher replied:
Hm I see

My issue with this is that it does not fit the intended design of the AnalysisStage

@RiccardoRossi I think you can evaluate best how this should be done

rubenzorrilla replied:
In this particular case what I'd do is to implement a do nothing `InitializeSolutionStep` and `Predict` and do everything, first for the thermal case and then for the structural one within the `SolveSolutionStep`.

AlejandroCornejo replied:
Done @rubenzorrilla

rubenzorrilla replied:
Please, comment the code explaining why this is done in such a way.

philbucher commented on applications/ConvectionDiffusionApplication/python_scripts/coupled_structural_thermal_solver.py:
```diff
@@ -141,19 +143,30 @@ def AdvanceInTime(self, current_time):
+        self.RemoveConvectiveVelocity()
 
-        return (fluid_is_converged and thermal_is_converged)
+        return (solid_is_converged)
```
```suggestion
        return solid_is_converged
```

philbucher commented on applications/StructuralMechanicsApplication/custom_utilities/constitutive_law_utilities.cpp:
```diff
@@ -336,15 +334,14 @@ void ConstitutiveLawUtilities<TVoigtSize>::CalculateElasticMatrixPlaneStrain(Mat
     if (rC.size1() != VoigtSize)
         rC.resize(VoigtSize, VoigtSize, false);
-    noalias(rC) = ZeroMatrix(VoigtSize, VoigtSize);
+    rC.clear();
```
are we sure this is the same?

AlejandroCornejo replied:
@loumalouomega

rubenzorrilla replied:
Indeed it is better. `ZeroMatrix` allocates a new one while `clear` just wipes everything inside.

philbucher commented on applications/StructuralMechanicsApplication/custom_utilities/constitutive_law_utilities.h:
```diff
@@ -243,16 +243,23 @@ class KRATOS_API(STRUCTURAL_MECHANICS_APPLICATION) ConstitutiveLawUtilities
-    static void CalculateElasticMatrix(MatrixType &C, ConstitutiveLaw::Parameters &rValues);
+    static void CalculateElasticMatrixPlaneStress(MatrixType &C, const double E, const double NU);
+    static void CalculateElasticMatrixPlaneStrain(MatrixType &C, const double E, const double N);
+    static void CalculateElasticMatrix(MatrixType &C, const double E, const double N);
```
```suggestion
    static void CalculateElasticMatrixPlaneStrain(MatrixType &C, const double E, const double NU);
    static void CalculateElasticMatrix(MatrixType &C, const double E, const double NU);
```

Or even better, use the full names (my recommendation)

AlejandroCornejo replied:
done

philbucher commented on applications/StructuralMechanicsApplication/custom_utilities/constitutive_law_utilities.h:
```diff
@@ -243,16 +243,23 @@ class KRATOS_API(STRUCTURAL_MECHANICS_APPLICATION) ConstitutiveLawUtilities
-    static void CalculateElasticMatrixPlaneStress(MatrixType &C, ConstitutiveLaw::Parameters &rValues);
-    static void CalculateElasticMatrixPlaneStrain(MatrixType &C, ConstitutiveLaw::Parameters &rValues);
-    static void CalculateElasticMatrix(MatrixType &C, ConstitutiveLaw::Parameters &rValues);
+    static void CalculateElasticMatrixPlaneStress(MatrixType &C, const double E, const double NU);
```
please add basic tests for those functions, given their importance

AlejandroCornejo committed:
corrected path

philbucher commented on applications/ConvectionDiffusionApplication/python_scripts/coupled_structural_thermal_solver.py:
can you please add a small full test case for this important feature?

AlejandroCornejo replied:
I've just added a case of 20 FE in which I am solving with the coupled thermomechanical solver including an accessor to modify the young modulus due to temperature and the addition of thermal strains. Basically is testing everything

AlejandroCornejo committed:
.

philbucher reviewed (commented):
some comments, otherwise :+1:

AlejandroCornejo committed:
comments

AlejandroCornejo committed:
trail whitespaces

AlejandroCornejo committed:
better stream

AlejandroCornejo committed:
adding complete test case

AlejandroCornejo committed:
tabs

AlejandroCornejo committed:
adding test to convection diffusion app

AlejandroCornejo committed:
implementing everything in calculateSolutionStep

AlejandroCornejo committed:
using pragma once

loumalouomega reviewed (changes requested):
I see a lot of code duplication in the code of the lasws, I whink we can simplify with a template (3D/PlaneStress/PlaneStrain) and define the base class in function of that template argument and the dimensions as well

loumalouomega commented on applications/ConstitutiveLawsApplication/custom_python/add_custom_utilities_to_python.h:
```diff
@@ -0,0 +1,42 @@
+
+
+#if !defined(KRATOS_ADD_UTILITIES_TO_PYTHON_H_INCLUDED )
+#define  KRATOS_ADD_UTILITIES_TO_PYTHON_H_INCLUDED
```
pragma once

loumalouomega commented on applications/ConstitutiveLawsApplication/custom_python/add_custom_utilities_to_python.h:
```diff
@@ -0,0 +1,42 @@
+namespace Kratos
+{
+
+namespace Python
```
Unify namespace: Kratos::Python

loumalouomega commented on applications/ConvectionDiffusionApplication/tests/thermo_mechanical_tests/thermo_mech_accessor/coupled_accessor_test.mdpa:
```diff
@@ -0,0 +1,314 @@
+Begin Properties 0
+End Properties
+Begin Nodes
+    1   0.0000000000   1.0000000000   0.0000000000
```
reduce saize of the mdpa removing modelpart data and change 0.0000000.. to 0.0

loumalouomega reviewed (changes requested):
My comments

AlejandroCornejo committed:
Merge branch 'master' into thermal-solver-mechanical-coupling

AlejandroCornejo committed:
reduce size of mdpa

AlejandroCornejo committed:
once

AlejandroCornejo committed:
unify namespaces

AlejandroCornejo committed:
removing unnecessary method

philbucher reviewed (commented):
tiny comment, then I will make a birthday present :)

AlejandroCornejo committed:
only dynamic in lowercase

philbucher reviewed (dismissed):
noice!

rubenzorrilla commented on applications/ConstitutiveLawsApplication/CMakeLists.txt:
```diff
@@ -16,6 +16,7 @@ file(GLOB_RECURSE KRATOS_CONSTITUTIVE_LAWS_APPLICATION_CORE
     ${CMAKE_CURRENT_SOURCE_DIR}/constitutive_laws_application.cpp
     ${CMAKE_CURRENT_SOURCE_DIR}/constitutive_laws_application_variables.cpp
     ${CMAKE_CURRENT_SOURCE_DIR}/custom_constitutive/*.cpp
+    ${CMAKE_CURRENT_SOURCE_DIR}/custom_constitutive_thermal/*.cpp
```
For the sake of consistency with the rest of applications, I'd suggest to put the new constitutive models together with their corresponding base classes within the `custom_constitutive` folder.

rubenzorrilla commented on applications/ConstitutiveLawsApplication/constitutive_laws_application.h:
```diff
@@ -104,6 +104,11 @@
 
 #include "custom_constitutive/small_strains/plastic_damage/associative_plastic_damage_model.h"
 
+// Thermal cl
```
```suggestion
// Thermal CL
```

rubenzorrilla commented on applications/ConstitutiveLawsApplication/custom_constitutive_thermal/small_strains/elastic/thermal_elastic_isotropic_3d.h:
```diff
@@ -0,0 +1,274 @@
+    double& GetReferenceTemperature()
+    {
+        return mReferenceTemperature;
+    }
```
This must be either `private` or `protected`.

rubenzorrilla commented on applications/ConstitutiveLawsApplication/custom_constitutive_thermal/small_strains/elastic/thermal_elastic_isotropic_3d.h:
```diff
@@ -0,0 +1,274 @@
+    void SetReferenceTemperature(const double ToRefTemperature)
+    {
+        mReferenceTemperature = ToRefTemperature;
+    }
```
Same.

rubenzorrilla commented on applications/ConstitutiveLawsApplication/custom_constitutive_thermal/small_strains/elastic/thermal_elastic_isotropic_3d.h:
```diff
@@ -0,0 +1,274 @@
+     */
+    void InitializeMaterial(const Properties &rMaterialProperties,
+                            const GeometryType &rElementGeometry,
+                            const Vector &rShapeFunctionsValues) override;
```
```suggestion
    void InitializeMaterial(
        const Properties &rMaterialProperties,
        const GeometryType &rElementGeometry,
        const Vector &rShapeFunctionsValues) override;
```

rubenzorrilla commented on applications/ConstitutiveLawsApplication/custom_constitutive_thermal/small_strains/elastic/thermal_elastic_isotropic_3d.h:
```diff
@@ -0,0 +1,274 @@
+
+// Project includes
+#include "custom_constitutive/elastic_isotropic_3d.h"
+#include "custom_utilities/advanced_constitutive_law_utilities.h"
```
I think you don't need this in the header but in the cpp.

rubenzorrilla commented on applications/ConstitutiveLawsApplication/custom_constitutive_thermal/small_strains/elastic/thermal_elastic_isotropic_3d.h:
```diff
@@ -0,0 +1,274 @@
+    ///@name Type Definitions
+    ///@{
+
+    typedef ElasticIsotropic3D BaseType;
```
```suggestion
    using BaseType = ElasticIsotropic3D ;
```

rubenzorrilla commented on applications/ConstitutiveLawsApplication/custom_constitutive_thermal/small_strains/elastic/thermal_linear_plane_strain.h:
```diff
@@ -0,0 +1,268 @@
+
+// Project includes
+#include "custom_constitutive/linear_plane_strain.h"
+#include "custom_utilities/advanced_constitutive_law_utilities.h"
```
To cpp.

rubenzorrilla commented on applications/ConstitutiveLawsApplication/custom_constitutive_thermal/small_strains/elastic/thermal_linear_plane_strain.h:
```diff
@@ -0,0 +1,268 @@
+    ///@name Type Definitions
+    ///@{
+
+    typedef LinearPlaneStrain BaseType;
```
Use `using`.

rubenzorrilla commented on applications/ConstitutiveLawsApplication/custom_constitutive_thermal/small_strains/elastic/thermal_linear_plane_strain.h:
```diff
@@ -0,0 +1,268 @@
+    ThermalLinearPlaneStrain() 
+    {
+
+    }
```
You can let the compiler to add it for you.

AlejandroCornejo replied:
Apparently in windows it fails to compile if I remove them...

rubenzorrilla replied:
And if you do it like this
````c++
ThermalLinearPlaneStrain() = default;
````

rubenzorrilla commented on applications/ConstitutiveLawsApplication/custom_constitutive_thermal/small_strains/elastic/thermal_linear_plane_strain.h:
```diff
@@ -0,0 +1,268 @@
+    void SetReferenceTemperature(const double ToRefTemperature)
+    {
+        mReferenceTemperature = ToRefTemperature;
+    }
```
Either `protected` or `private`. Besides doxygen documentation is missing.

rubenzorrilla commented on applications/ConstitutiveLawsApplication/custom_constitutive_thermal/small_strains/elastic/thermal_linear_plane_strain.h:
```diff
@@ -0,0 +1,268 @@
+    void InitializeMaterial(const Properties &rMaterialProperties,
+                            const GeometryType &rElementGeometry,
+                            const Vector &rShapeFunctionsValues) override;
+
```
Make indentation consistent with the rest of the file.

rubenzorrilla commented on applications/ConstitutiveLawsApplication/custom_constitutive_thermal/small_strains/elastic/thermal_linear_plane_stress.h:
```diff
@@ -0,0 +1,277 @@
+
+// Project includes
+#include "custom_constitutive/linear_plane_stress.h"
+#include "custom_utilities/advanced_constitutive_law_utilities.h"
```
To cpp.

rubenzorrilla commented on applications/ConstitutiveLawsApplication/custom_constitutive_thermal/small_strains/elastic/thermal_linear_plane_stress.h:
```diff
@@ -0,0 +1,277 @@
+    ///@name Type Definitions
+    ///@{
+
+    typedef LinearPlaneStress BaseType;
```
Use `using`.

rubenzorrilla commented on applications/ConstitutiveLawsApplication/custom_constitutive_thermal/small_strains/elastic/thermal_linear_plane_stress.h:
```diff
@@ -0,0 +1,277 @@
+    ThermalLinearPlaneStress() 
+    {
+
+    }
```
The compiler should automatically add this one.

rubenzorrilla commented on applications/ConstitutiveLawsApplication/custom_constitutive_thermal/small_strains/elastic/thermal_linear_plane_stress.h:
```diff
@@ -0,0 +1,277 @@
+     */
+    void InitializeMaterial(const Properties &rMaterialProperties,
+                            const GeometryType &rElementGeometry,
+                            const Vector &rShapeFunctionsValues) override;
```
Make indentation consistent.

rubenzorrilla commented on applications/ConstitutiveLawsApplication/custom_constitutive_thermal/small_strains/elastic/thermal_linear_plane_stress.h:
```diff
@@ -0,0 +1,277 @@
+    void SetReferenceTemperature(const double ToRefTemperature)
+    {
+        mReferenceTemperature = ToRefTemperature;
+    }
```
Either `protected` or `private`.

rubenzorrilla commented on applications/ConstitutiveLawsApplication/custom_python/add_custom_constitutive_laws_to_python.cpp:
```diff
@@ -97,6 +97,11 @@
 
 #include "custom_constitutive/small_strains/plastic_damage/associative_plastic_damage_model.h"
 
+// Thermal cl
```
```suggestion
// Thermal CL
```

rubenzorrilla commented on applications/ConstitutiveLawsApplication/custom_python/add_custom_utilities_to_python.cpp:
```diff
@@ -0,0 +1,35 @@
+
+
+namespace Kratos {
+namespace Python {
```
```suggestion
namespace Kratos::Python {
```

rubenzorrilla commented on applications/ConstitutiveLawsApplication/custom_python/add_custom_utilities_to_python.cpp:
```diff
@@ -0,0 +1,35 @@
+}
+
+}  // namespace Python.
+} // Namespace Kratos
```
```suggestion
} // namespace Kratos::Python
```

rubenzorrilla commented on applications/ConstitutiveLawsApplication/custom_python/add_custom_utilities_to_python.h:
```diff
@@ -0,0 +1,34 @@
+{
+
+void  AddCustomUtilitiesToPython(pybind11::module& m);
+
+}  // namespace Kratos.
```
```suggestion

}  // namespace Kratos::Python.
```

rubenzorrilla commented on applications/ConstitutiveLawsApplication/custom_utilities/advanced_constitutive_law_utilities.cpp:
```diff
@@ -759,6 +759,51 @@ void AdvancedConstitutiveLawUtilities<TVoigtSize>::CalculateRotationOperator(
+    unsigned int step
+    )
+{
+    const GeometryType& r_geometry = rParameters.GetElementGeometry();
```
```suggestion
    const auto& r_geometry = rParameters.GetElementGeometry();
```

rubenzorrilla commented on applications/ConstitutiveLawsApplication/custom_utilities/advanced_constitutive_law_utilities.h:
```diff
@@ -476,5 +477,22 @@ class KRATOS_API(CONSTITUTIVE_LAWS_APPLICATION) AdvancedConstitutiveLawUtilities
     static double MacaullyBrackets(
         const double Number);
```
```suggestion
    static double MacaullyBrackets(const double Number);
```
Style guide (one argument methods must be in one line).

rubenzorrilla commented on applications/ConstitutiveLawsApplication/tests/cpp_tests/test_thermal_constitutive_laws.cpp:
```diff
@@ -0,0 +1,387 @@
+{
+
+namespace Testing
+{
```
```suggestion
namespace Kratos::Testing
{
```

rubenzorrilla commented on applications/ConstitutiveLawsApplication/tests/cpp_tests/test_thermal_constitutive_laws.cpp:
```diff
@@ -0,0 +1,387 @@
+    }
+
+} // namespace Testing
+} // namespace Kratos.
```
```suggestion
} // namespace Kratos::Testing.
```

rubenzorrilla commented on applications/ConstitutiveLawsApplication/tests/cpp_tests/test_thermal_constitutive_laws.cpp:
```diff
@@ -0,0 +1,387 @@
+// KRATOS ___                _   _ _         _   _             __                       _
```
There is a a lot of copy&paste code in this test. I suggest using moving this to an auxiliary function (you can place this new one in an unnamed namespace).

rubenzorrilla commented on applications/ConvectionDiffusionApplication/python_scripts/coupled_structural_thermal_solver.py:
```diff
@@ -141,19 +143,34 @@ def AdvanceInTime(self, current_time):
 
-        return (fluid_is_converged and thermal_is_converged)
+        KratosMultiphysics.Logger.PrintInfo("\t" + "Solving THERMAL part...")
+        self.thermal_solver.SolveSolutionStep()
```
```suggestion
        thermal_is_converged = self.thermal_solver.SolveSolutionStep()
```

rubenzorrilla commented on applications/ConvectionDiffusionApplication/python_scripts/coupled_structural_thermal_solver.py:
```diff
@@ -141,19 +143,34 @@ def AdvanceInTime(self, current_time):
+
+        self.RemoveConvectiveVelocity()
+
+        return solid_is_converged
```
```suggestion
        return solid_is_converged and thermal_is_converged
```

rubenzorrilla commented on applications/ConvectionDiffusionApplication/python_scripts/coupled_structural_thermal_solver.py:
```diff
@@ -141,19 +143,34 @@ def AdvanceInTime(self, current_time):
     def FinalizeSolutionStep(self):
         self.structural_solver.FinalizeSolutionStep()
         self.thermal_solver.FinalizeSolutionStep()
+        KratosMultiphysics.Logger.PrintInfo("\n")
```
```suggestion
```

rubenzorrilla commented on applications/ConvectionDiffusionApplication/tests/thermo_mechanical_tests/thermo_mech_accessor/ConvectionDiffusionMaterials.json:
```diff
@@ -0,0 +1,14 @@
+                "CONDUCTIVITY"  : 50.0,
+                "SPECIFIC_HEAT" : 460.0
+            },
+            "Tables"    : null
```
```suggestion
            }
```

rubenzorrilla commented on applications/ConvectionDiffusionApplication/tests/convection_diffusion_test_factory.py:
```diff
@@ -79,4 +79,8 @@ class BasicDiffusionStationaryTest(ConvectionDiffusionTestFactory):
+
+@KratosUnittest.skipIfApplicationsNotAvailable("StructuralMechanicsApplication")
+class SimpleThermoMechanicalTableAccessorTest(ConvectionDiffusionTestFactory):
+    file_name = "thermo_mechanical_tests/thermo_mech_accessor/coupled_accessor_test"
```
```suggestion
    file_name = "thermo_mechanical_tests/thermo_mechanical_accessor/coupled_accessor_test"
```
Besides I don't think the subfolder is required at all.

rubenzorrilla commented on applications/ConvectionDiffusionApplication/tests/thermo_mechanical_tests/thermo_mech_accessor/coupled_accessor_test.mdpa:
```diff
@@ -0,0 +1,310 @@
+Begin Properties 0
```
Can't we do the same with a much coarse mesh. I think that very few elements are sufficient in this case).

rubenzorrilla commented on applications/ConvectionDiffusionApplication/tests/thermo_mechanical_tests/thermo_mech_accessor/coupled_accessor_test_parameters.json:
```diff
@@ -0,0 +1,191 @@
+        "problem_name"  : "coupled_problem",
+        "parallel_type" : "OpenMP",
+        "start_time"    : 0.0,
+        "end_time"      : 50,
```
I think that running a single time step is sufficient in this case.

AlejandroCornejo replied:
well... the heat flow takes some time to evolve. I'll reduce the simulation time anyways

rubenzorrilla replied:
I don't think this matters at all. At the end of the day we want to check that the values are always the same. Alternatively, you can set a higher conductivity.

rubenzorrilla commented on applications/ConvectionDiffusionApplication/tests/thermo_mechanical_tests/thermo_mech_accessor/coupled_accessor_test_parameters.json:
```diff
@@ -0,0 +1,191 @@
+        "domain_size"             : 2,
+        "echo_level"              : 0,
+        "structural_solver_settings" : {
+            "solver_type"                     : "Static",
```
This test doesn't cover the convective term removal.

rubenzorrilla commented on applications/ConvectionDiffusionApplication/tests/thermo_mechanical_tests/thermo_mech_accessor/coupled_accessor_test_parameters.json:
```diff
@@ -0,0 +1,191 @@
+                    },
+                    "point_data_configuration"  : []
+                }
+            }
```
GiD output must be optional (only for eventual debugging).

rubenzorrilla commented on applications/StructuralMechanicsApplication/custom_utilities/constitutive_law_utilities.cpp:
```diff
@@ -286,20 +286,20 @@ void ConstitutiveLawUtilities<TVoigtSize>::CalculateRotationOperatorVoigt(
-    const double E = r_material_properties[YOUNG_MODULUS];
-    const double NU = r_material_properties[POISSON_RATIO];
-
     if (rC.size1() != VoigtSize)
```
```suggestion
    if (rC.size1() != VoigtSize || rC.size2() != VoigtSize)
```

rubenzorrilla commented on applications/StructuralMechanicsApplication/custom_utilities/constitutive_law_utilities.cpp:
```diff
@@ -311,20 +311,19 @@ void ConstitutiveLawUtilities<TVoigtSize>::CalculateElasticMatrixPlaneStress(Mat
-    const double E = r_material_properties[YOUNG_MODULUS];
-    const double NU = r_material_properties[POISSON_RATIO];
-
     if (rC.size1() != VoigtSize)
```
```suggestion
    if (rC.size1() != VoigtSize || rC.size2() != VoigtSize)
```

rubenzorrilla commented on applications/StructuralMechanicsApplication/custom_utilities/constitutive_law_utilities.cpp:
```diff
@@ -336,20 +335,19 @@ void ConstitutiveLawUtilities<TVoigtSize>::CalculateElasticMatrixPlaneStrain(Mat
-    const double E  = r_material_properties[YOUNG_MODULUS];
-    const double NU = r_material_properties[POISSON_RATIO];
-
     if (rC.size1() != VoigtSize)
```
```suggestion
    if (rC.size1() != VoigtSize || rC.size2() != VoigtSize)
```

rubenzorrilla commented on applications/StructuralMechanicsApplication/custom_utilities/constitutive_law_utilities.h:
```diff
@@ -243,16 +243,23 @@ class KRATOS_API(STRUCTURAL_MECHANICS_APPLICATION) ConstitutiveLawUtilities
+    */
+    static void CalculatePK2StressFromStrain(ConstitutiveLaw::StressVectorType& rStressVector, const ConstitutiveLaw::StrainVectorType &rStrainVector, const double YoungModulus, const double PoissonRatio);
+    static void CalculatePK2StressFromStrainPlaneStress(ConstitutiveLaw::StressVectorType& rStressVector, const ConstitutiveLaw::StrainVectorType &rStrainVector, const double YoungModulus, const double PoissonRatio);
+    static void CalculatePK2StressFromStrainPlaneStrain(ConstitutiveLaw::StressVectorType& rStressVector, const ConstitutiveLaw::StrainVectorType &rStrainVector, const double YoungModulus, const double PoissonRatio);
```
Please follow style guide and add Doxygen documentation.

AlejandroCornejo committed:
moving methods to private and add doxigen

AlejandroCornejo committed:
consistent style

AlejandroCornejo committed:
includes to cpp

AlejandroCornejo committed:
not typedef anymore

AlejandroCornejo committed:
collapsing namespaces

AlejandroCornejo committed:
review advanced CL utilities

AlejandroCornejo committed:
avoid repetition int he cpp test

AlejandroCornejo committed:
adding conv in thermal solve

AlejandroCornejo committed:
remove print

AlejandroCornejo committed:
reduced size mdpa

AlejandroCornejo committed:
remove gid and vtk print settings in test

AlejandroCornejo committed:
adding extra check for C

AlejandroCornejo committed:
correcting header of utilities and style

AlejandroCornejo committed:
chaging name test directory

AlejandroCornejo committed:
moving CLs to its approppriate folder

AlejandroCornejo committed:
update directory in json test

AlejandroCornejo committed:
now we run only 5 steps intead of 50

loumalouomega commented on applications/ConstitutiveLawsApplication/constitutive_laws_application.h:
```diff
@@ -564,6 +569,12 @@ class KRATOS_API(CONSTITUTIVE_LAWS_APPLICATION) KratosConstitutiveLawsApplicatio
+    const ThermalLinearPlaneStrain mThermalLinearPlaneStrain;
+    const ThermalLinearPlaneStress mThermalLinearPlaneStress;
+
+
```
```suggestion

```

loumalouomega committed:
Minoor suggestions

Co-authored-by: Rubén Zorrilla <rubenzorrillamartinez@hotmail.com>

loumalouomega reviewed (commented):
My suggestion of templated class is not addressed

AlejandroCornejo committed:
Merge branch 'master' into thermal-solver-mechanical-coupling

AlejandroCornejo committed:
ading a virtual method to the thermal one 3d

AlejandroCornejo committed:
adapting plane strain to thermal

AlejandroCornejo committed:
moving stuff to utilities and avoid repetition

AlejandroCornejo committed:
with const expr

AlejandroCornejo committed:
corrected

AlejandroCornejo committed:
unused GetValue removed

AlejandroCornejo committed:
Revert "unused GetValue removed"

This reverts commit 22e35f3c4ffb3662ebf847af4b1340af057a160b.

AlejandroCornejo committed:
iso dam ok

AlejandroCornejo committed:
h of isodam

AlejandroCornejo committed:
plane strain ok

AlejandroCornejo committed:
adding dimension to thermal plane strain

AlejandroCornejo committed:
updating plane stress

loumalouomega reviewed (dismissed):
Removing my block

AlejandroCornejo committed:
more generality to avoid breaking test in SMApp

rubenzorrilla reviewed (changes requested):
The tests subsubfolder are still in there. Besides, the subfolder is at the level of the files of another test, something that is way misleading.

Please, either wrap the already existent test files in a new folder or simply put the `thermo_mechanical_accessor` at the same level.

AlejandroCornejo committed:
moving the test to a subfolder to be consistent

rubenzorrilla reviewed (approved):
All my concerns have been addressed. Good job.
